Creating a Python cheat sheet

NoSQLKnowHow opened the pull request:
More cheat sheets to come.

NoSQLKnowHow committed:
Create momento-cache-python-cheat-sheet.md

NoSQLKnowHow committed:
Update momento-cache-python-cheat-sheet.md

NoSQLKnowHow committed:
added more functions and the category json

malandis commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
```diff
@@ -0,0 +1,107 @@
+
+  def client():
+    momento_auth_token = CredentialProvider.from_environment_variable('MOMENTO_AUTH_TOKEN')
+    ttl  = timedelta(int(os.getenv('MOMENTO_TTL_SECONDS')))
```
Recommend being explicit with the timedelta constructor units and also offering a default TTL if the env var doesn't exist
```suggestion
    ttl  = timedelta(seconds=int(os.getenv('MOMENTO_TTL_SECONDS', '60')))
```

malandis commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,107 @@
+            print(f"Error incrementing value: {error.message}")
+        case _:
+            print("Unreachable")
+```
````
Maybe demo this idea?
```suggestion
A simple example of doing a set operation. In the client.set call, you could pass a fourth value, the TTL and this would override the default TTL value set with the client connection object.
```python
  def set(cache_name, key, value, ttl=None):
    with MomentoCounter.client() as client:
      resp = client.set(cache_name, key, value, ttl)
      match resp:
        case CacheSet.Success():
            pass
        case CacheSet.Error() as error:
            print(f"Error incrementing value: {error.message}")
        case _:
            print("Unreachable")
```
```

malandis commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,107 @@
+This is an example of a simple read operation to get an item from a cache.
+```python
+  def get(cache_name, key):
+    cache_name = os.getenv('MOMENTO_CACHE_NAME')
````
We read the cache name from this example function but not in the others. Was this left by mistake?

NoSQLKnowHow replied:
Fixed

malandis commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,107 @@
+```python
+  def incr(cache_name, key, value:int = 1):
+    with MomentoCounter.client() as client:
+      resp = client.increment(cache_name,key,value)
````
My inner python formatter 🙈 

```suggestion
      resp = client.increment(cache_name, key, value)
```

NoSQLKnowHow replied:
Fixed

malandis reviewed (commented):
First pass -- I like the snippets. Mostly just some style recs

NoSQLKnowHow committed:
Update docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md

Co-authored-by: Michael Landis <malandis@gmail.com>

NoSQLKnowHow committed:
Update docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md

Co-authored-by: Michael Landis <malandis@gmail.com>

NoSQLKnowHow committed:
Update momento-cache-python-cheat-sheet.md

NoSQLKnowHow committed:
Merge branch 'python-cheat-sheet' of https://github.com/momentohq/public-dev-docs into python-cheat-sheet

malandis commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
```diff
@@ -0,0 +1,105 @@
+
+import os
+
+  def client():
```
```suggestion
def client() -> CacheClient:
```

NoSQLKnowHow replied:
A note will be added to the bottom on how to do this if you want to reuse the client connection object.

malandis commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,105 @@
+## Create a new cache in Momento Serverless Cache
+Use this function to create a new cache in your account
+```python
+  def create_cache(cache_name):
````
```suggestion
  def create_cache(cache_name) -> None:
```

malandis commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,105 @@
+Use this function to create a new cache in your account
+```python
+  def create_cache(cache_name):
+    with MomentoCounter.client() as client:
````
I don't see where `MomentoCounter` is defined here. Did you mean just `client()`?

NoSQLKnowHow replied:
MomentoCounter is the name of the class these functions are a part of...and yeah, I omitted that. i think I need to put a link to the entire class file to refer to.

malandis reviewed (commented):
Couple other style points.

Another thing I noticed is that each of the cache operation functions instantiates a new cache client (`with ...client() as client:`). This will means the application will reconnect on each operation. I recommend passing the client as an argument to the operations to make clear they should re-use the cache client.

NoSQLKnowHow committed:
Update docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md

Co-authored-by: Michael Landis <malandis@gmail.com>

NoSQLKnowHow committed:
Updated cheat sheet and added entire file

Updated the cheat sheet to change name of class and added entire class file to the /static directory

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,115 @@
+```
+
+## Create a new cache in Momento Serverless Cache
+Use this function to create a new cache in your account
````
This one does not have a period at the end, but others do

NoSQLKnowHow replied:
Good catch. fixed.

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,115 @@
+If you combine all of the functions on this page into one python file, you'd have a central collection of functions you can import and call from other python code. In addition, if you are putting this into production, you might look to replace the print() calls with ones using the logging library (`import logging`) in python.
+
+## Import libraries and connect to return a CacheClient object
+```python
````
Do you think you need a short description here to be consistent with the others?

NoSQLKnowHow replied:
Good idea. Added one.

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,115 @@
+```
+
+## Get list of existing caches in your account
+In this example, we use the client function above to get a client object and then use it to list all of the cache's for this account.
````
You have `cache's` in here, but it should just be `caches`

NoSQLKnowHow replied:
fixed

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
```diff
@@ -0,0 +1,115 @@
+                for cache_info in success.caches:
+                    print(f"- {cache_info.name!r}")
+            case ListCaches.Error() as error:
+                print(f"Error creating cache: {error.message}")
```
This error message is for the wrong call

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
```diff
@@ -0,0 +1,115 @@
+        case CreateCache.Success():
+            print("Cache created.")
+        case CreateCache.Error() as error:
+            print(f"Error incrementing value: {error.message}")
```
This error message isn't right

NoSQLKnowHow replied:
Me copy and pasting. fixed.

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,115 @@
+        print("")
+```
+## Write an item to a cache
+A simple example of doing a set operation. In the client.set call, you could pass a fourth value, the TTL and this would override the default TTL value set with the client connection object.
````
You need a comma after "the TTL" in here

NoSQLKnowHow replied:
rewrote this.

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
```diff
@@ -0,0 +1,115 @@
+        case CacheSet.Success():
+            pass
+        case CacheSet.Error() as error:
+            print(f"Error incrementing value: {error.message}")
```
This error message is wrong

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
```diff
@@ -0,0 +1,115 @@
+            print(f"Error incrementing value: {error.message}")
+        case _:
+            print("Unreachable")
+
```
You forgot to close the code block here (missing ```)

NoSQLKnowHow replied:
Fixed

NoSQLKnowHow committed:
reordered some text

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
```diff
@@ -0,0 +1,115 @@
+    with MomentoBasics.client() as client:
+      resp = client.set(cache_name, key, value, ttl)
+      match resp:
+        case CacheSet.Success():
```
Is this right? it's very different than below where you're using
```python
isinstance(resp, CacheGet.Hit)
```

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,115 @@
+        raise resp.inner_exception
+```
+## Increment the value of an item in a cache
+An example of incrementing a value in a key. You can pass in any number
````
Missing a period at the end of this description. Should also indicate increment uses integers and does not accept decimals

NoSQLKnowHow replied:
Fixed the period and added more text to clarify about the integer.

NoSQLKnowHow committed:
changed notes to tip and info

allenheltondev commented on docs/develop/guides/cheat-sheets/momento-cache-python-cheat-sheet.md:
````diff
@@ -0,0 +1,115 @@
+def client() -> CacheClient:
+```
+
+The modify the other function definitions to accept a CacheClient object as an argument, but also not instantiate its own object.
````
Typo here. It leads with "The modify" but probably was intended to be "Then, modify"

allenheltondev commented on static/code/cheat-sheets/MomentoBasics.py:
```diff
@@ -0,0 +1,73 @@
+        case CreateCache.Success():
+            print("Cache created.")
+        case CreateCache.Error() as error:
+            print(f"Error incrementing value: {error.message}")
```
These error messages need to be updated to match the ones in the cheat sheet

NoSQLKnowHow committed:
fixed more things after Allen's review.

NoSQLKnowHow committed:
Fixed minor text.

allenheltondev commented on static/code/cheat-sheets/MomentoBasics.py:
```diff
@@ -0,0 +1,73 @@
+  def get(cache_name, key):
+    with MomentoBasics.client() as client:
+      resp = client.get(cache_name, key)
+      if isinstance(resp, CacheGet.Hit):
```
This one is still in that other format and doesn't match the example from the cheat sheet

NoSQLKnowHow replied:
*sigh* fixed.

NoSQLKnowHow committed:
Update MomentoBasics.py
